Do not rebuild Routes when a new opaque Route is pushed on top

goderbauer opened the pull request:
Also known as: Don't rebuild OverlayEntries when one changes opaqueness.

## Description

Prior to this change we were re-arranging the tree when an OverlayEntry went from onstage to offstage (due to opaqueness changes) and vice versa:

https://github.com/flutter/flutter/blob/4f9b6cf01799a60ec6c4cc55c614a0e76ca80a97/packages/flutter/lib/src/widgets/overlay.dart#L461-L467

Since the OverlayEntries have a global key their corresponding element was deactivated at the old location inside the Stack and re-activated at the new location outside the Stack. This caused all StatefulWidgets within that subtree to rebuild because their ancestor had changed.

This change avoids the rebuild by not re-arranging the tree. Instead, both onstage and offstage OverlayEntries are children of the _Theatre, so they don't have to be moved around in the tree. The RenderObject of the _Theatre implements a special version of a Stack that just ignores the offstage children during rendering and layout. (Since the offstage children are always all at the beginning of the child list, this is implemented by providing an integer skipCount indicating the number of children to skip.)

The unnecessary rebuilds were most commonly felt when pushing new Routes because Navigator uses an Overlay internally. 

## Related Issues

Fixes https://github.com/flutter/flutter/issues/45797.

## Tests

I added the following tests:

* OverlayEntries do not rebuild when opaque entry is added
* OverlayEntries do not rebuild when opaqueness changes
* entries below opaque entries are ignored for hit testing
* Semantics of entries below opaque entries are ignored
* Can used Positioned within OverlayEntry
* Pushing opaque Route does not rebuild routes below

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [x] I wrote a migration guide: https://github.com/flutter/website/pull/3581

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

goderbauer committed:
RenderTheatre implements Stack

goderbauer committed:
more tests

goderbauer committed:
self review

goderbauer committed:
more tests

goderbauer committed:
another test

goderbauer commented:
@Hixie Even though I had to update framework tests for this change I would like to handle it as a non-breaking change. Can you grant an exception? 

The changes to the tests were to update error messages and error-message-like output. In one instance I had to change a type due to the changed internals of _Theatre.

dnfield commented on packages/flutter/lib/src/widgets/overlay.dart:
```diff
@@ -442,28 +451,32 @@ class OverlayState extends State<Overlay> with TickerProviderStateMixin {
-        children: onstageChildren.reversed.toList(growable: false),
-      ),
-      offstage: offstageChildren,
+      skipCount: children.length - onstageCount,
```
This seems like it would fail if the logic above added children in an order like [offstage, onstage, offstage, offstage] for some reason.

If we think that's invalid, we should probably assert it. If it's actually valid we need different logic here.

goderbauer replied:
Doing something like [offstage, onstage, offstage, offstage] is not supported by the overlay. The top x entries are always onstage, everything else is either completely ignored or - if maintainState is true - rendered offstage.

Not sure what we should assert here. The assert would basically duplicate the algorithm above.

dnfield commented on packages/flutter/lib/src/widgets/overlay.dart:
```diff
@@ -476,223 +489,319 @@ class OverlayState extends State<Overlay> with TickerProviderStateMixin {
+  @override
+  void debugVisitOnstageChildren(ElementVisitor visitor) {
+    assert(children.length >= widget.skipCount);
+    children.skip(widget.skipCount).forEach(visitor);
```
Similar concern - I think we at least want an assert here that all the widgets we're skipping are in fact offstage.

goderbauer replied:
It's the other way around, though: The fact that we are skipping them makes them offstage. There's no other marker that marks them as offstage that we could assert on.

dnfield commented on packages/flutter/lib/src/widgets/overlay.dart:
```diff
@@ -476,223 +489,319 @@ class OverlayState extends State<Overlay> with TickerProviderStateMixin {
+  int get skipCount => _skipCount;
+  int _skipCount;
+  set skipCount(int value) {
+    assert(value != null);
```
assert that the children we're skipping are offstage?

goderbauer replied:
Same answer as above: The fact that we are skipping them makes them offstage. Not the other way around.

dnfield commented on packages/flutter/lib/src/widgets/overlay.dart:
```diff
@@ -476,223 +489,319 @@ class OverlayState extends State<Overlay> with TickerProviderStateMixin {
+      );
+      if (isHit)
+        return true;
+      child = childParentData.previousSibling;
```
assert the previous sibling is onstage?

goderbauer replied:
Same.

dnfield reviewed (commented):
FWIW, I agree changing the tests shouldn't be considered breaking here.

I may be missing something, but the skipCount seems sketchy to me.  I think we shoudl at least be asserting that the children we think are offstage are in fact offstage, and vice-versa.

Is there ever a case where an opaque route should still layout other routes?  Could we have a route that doesn't cover the whole view?

goderbauer commented:
> I may be missing something, but the skipCount seems sketchy to me. I think we shoudl at least be asserting that the children we think are offstage are in fact offstage, and vice-versa.

The definition of the (private) _Theatre widget is: "give me a list of children and a number n. I'll not layout/render the first n children (effectively putting them offstage)."

The number n defines what is going to be offstage. There's no other mechanism to tell of a child should be offstage or not...

dnfield reviewed (approved):
LGTM

Hixie commented:
I haven't reviewed the test of the patch (I'm on my phone) but the test changes call be considered exempt from the breaking change policy as they are only testing internals.

goderbauer commented:
This change positively affected some of our Gallery transition benchmarks:

<img width="270" alt="Screen Shot 2020-01-22 at 9 24 29 AM" src="https://user-images.githubusercontent.com/1227763/72917615-0369a580-3cf9-11ea-8401-d8ae70e11a81.png">

* dropped from 63 frames to 50 (minus 21%)

<img width="279" alt="Screen Shot 2020-01-22 at 9 24 23 AM" src="https://user-images.githubusercontent.com/1227763/72917616-0369a580-3cf9-11ea-9e72-000b0af32435.png">

* dropped from 50.45ms to 36.64ms (minus 27%)

<img width="297" alt="Screen Shot 2020-01-22 at 9 21 54 AM" src="https://user-images.githubusercontent.com/1227763/72917429-acfc6700-3cf8-11ea-8e58-ddf53c09dac8.png">

* dropped from 7.38ms to 6.13ms (minus 17%)

dnfield commented:
Wow I'm so used to seeing posts like that linked to a today bug due to a regression.  Nice!

dkwingsmt commented:
If we were to reland this PR, make sure to address the issue mentioned in https://github.com/flutter/flutter/pull/49318, where the order between multiple dependencies are not consistent. They might change due to an irrelevant change, or be different between platforms.

shamilovtim commented:
Hello, is this released?
